Stop filling in defaults in Python binding layer.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33759 Stop filling in defaults in Python binding layer.**
* #33825 Turn on strict dtype checking for test_torch.py
* #33510 Correctly preserve "not set anywhere" TensorOptions when merging.
* #33505 Switch empty_like to use merge_in to process TensorOptions.

Now that C++ API has TensorOptions merging bugs fixed, we can now
just have the Python binding code accurately report when a kwarg
wasn't set, instead of filling in the defaults itself (and getting
it wrong, sometimes.)

Along the way, I made the python_arg_parser store a Layout directly
instead of a THPLayout, also nicely simplifying some more of the logic.

Differential Revision: [D20096635](https://our.internmc.facebook.com/intern/diff/D20096635)

ezyang committed:
Stop filling in defaults in Python binding layer.

Now that C++ API has TensorOptions merging bugs fixed, we can now
just have the Python binding code accurately report when a kwarg
wasn't set, instead of filling in the defaults itself (and getting
it wrong, sometimes.)

Along the way, I made the python_arg_parser store a Layout directly
instead of a THPLayout, also nicely simplifying some more of the logic.

[ghstack-poisoned]

bhosmer reviewed (approved):
This looks great as far as it goes, but wouldn't we also want to cash in the fix by deleting https://github.com/pytorch/pytorch/pull/33759/files#diff-78707e9331c95eed3d8f7d6186b0eb28L346-L347? Accepting on the premise that there's a reason you didn't :) 

[edit: that line highlight link doesn't show automatically, it's this
```

    return (scalartype == at::ScalarType::Undefined) ?
            torch::tensors::get_default_scalar_type() : scalartype;
 ```

torch/csrc/utils/python_arg_parser.h 348-349

ezyang commented:
I think the non-optional python arg parser methods can be deleted entirely, but that got big enough that I'd put it in a separate PR.

ezyang committed:
Update on "Stop filling in defaults in Python binding layer."


Now that C++ API has TensorOptions merging bugs fixed, we can now
just have the Python binding code accurately report when a kwarg
wasn't set, instead of filling in the defaults itself (and getting
it wrong, sometimes.)

Along the way, I made the python_arg_parser store a Layout directly
instead of a THPLayout, also nicely simplifying some more of the logic.

Differential Revision: [D20096635](https://our.internmc.facebook.com/intern/diff/D20096635)

[ghstack-poisoned]

ezyang committed:
Update on "Stop filling in defaults in Python binding layer."


Now that C++ API has TensorOptions merging bugs fixed, we can now
just have the Python binding code accurately report when a kwarg
wasn't set, instead of filling in the defaults itself (and getting
it wrong, sometimes.)

Along the way, I made the python_arg_parser store a Layout directly
instead of a THPLayout, also nicely simplifying some more of the logic.

Differential Revision: [D20096635](https://our.internmc.facebook.com/intern/diff/D20096635)

[ghstack-poisoned]

ezyang committed:
Update on "Stop filling in defaults in Python binding layer."


Now that C++ API has TensorOptions merging bugs fixed, we can now
just have the Python binding code accurately report when a kwarg
wasn't set, instead of filling in the defaults itself (and getting
it wrong, sometimes.)

Along the way, I made the python_arg_parser store a Layout directly
instead of a THPLayout, also nicely simplifying some more of the logic.

Differential Revision: [D20096635](https://our.internmc.facebook.com/intern/diff/D20096635)

[ghstack-poisoned]

ezyang committed:
Update on "Stop filling in defaults in Python binding layer."


Now that C++ API has TensorOptions merging bugs fixed, we can now
just have the Python binding code accurately report when a kwarg
wasn't set, instead of filling in the defaults itself (and getting
it wrong, sometimes.)

Along the way, I made the python_arg_parser store a Layout directly
instead of a THPLayout, also nicely simplifying some more of the logic.

Differential Revision: [D20096635](https://our.internmc.facebook.com/intern/diff/D20096635)

[ghstack-poisoned]

ezyang commented:
This was incorrectly closed; the new PR is at https://github.com/pytorch/pytorch/pull/33759
